### PR TITLE
create milestone5 branch

### PR DESCRIPTION
**Pull request by Igors78**
Milestone friendships v.1 
- Create a model with associations and all requested features for friendships.

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -1,41 +1,45 @@
...
-      </div>
+    </div>
     <% end %>
     <% if alert.present? %>
-      <div class="alert">
+    <div class="alert">
         <p> <%= alert %> </p>
-      </div>
+    </div>
     <% end %>
```
Comment (lekegitrepo) [inline app/views/layouts/application.html.erb:39]: - **_Remove logic from the view_**

File: app/views/users/_user.html.erb [hunk 2]
```diff
@@ -1,6 +1,23 @@
...
+                                class: "badge"%>
+        <% else %>
+        <%= link_to 'Befriend',
+                             { controller: "friendships",
+                               action: "create_friendship",
+                               user_id: current_user.id,
+                               friend_id: user.id },
+        class: 'badge badge-secondary' %>
+        <% end %>
+        <% end %>
```
Comment (lekegitrepo) [inline app/views/users/_user.html.erb:21]: - **_Remove logic from the view_**

File: app/views/users/show.html.erb [hunk 3]
```diff
@@ -1,9 +1,20 @@
...
+    <% if current_user == @user && @user.friend_requests_received.any?%>
+    <% @user.friend_requests_received.each do |request| %>
+    <p>Friendship requested by:<%=  "#{request.user.name}"%></p>
+    <%= link_to "Accept request",
+                                { controller: "friendships",
+                                  action: "accept_request",
+                                  user_id: request.user.id,
+                                  friend_id: request.friend.id },
+        class: 'badge badge-primary' %>
+    <% end %>
```
Comment (lekegitrepo) [inline app/views/users/show.html.erb:12]: - **_Remove logic from the view_**

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Changes Required! :recycle:
### Hi @Igors78 and @CornerstoneII :clap: :clap:
#### You have done a good job. The model association is well implemented :muscle:
#### However, some requirements are missing/not implemented correctly. Check my comments below :point_down:

#### 1) LOGIC IN VIEW:
- **There should be NO application logic in the views. Everything connected with data processing should be encapsulated in models and used in controllers.**
  * [x] **_View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb)_**
  * [x] **_There should be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it._**

**For more details check out the following :point_down: resources**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

#### _Reviewed by Temitayo Adeleke_

**Review by Temmarie — CHANGES_REQUESTED** [review]
## Changes Required! :recycle:

### Hi @Igors78  ! :wave: :wave:

#### You have done well so far :clap: However, there are changes required for you to make in order for this project to be approved. Kindly work on them and submit them for another code review.


#### Behaviour

- [ ] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- [ ] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- [x] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- You can do this by creating another view or have accept and decline buttons next to the name of the user who sent the friend request in the ALL USERS tab

- [ ] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- The timeline should only have the login user's post and his or her friend's post so let's create these validations for the views.

#### Test

- [ ] Remove all empty test files.

### That's it, team! :raised_hands: 


__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Comment by ioilmio** [comment]
# Hello @Igors78 👋🏽 
## I wasn't able to review your project because you requested a Code Review for the wrong milestone.
## Please submit again selecting the correct milestone.

## Happy coding 💻

**Review by Temmarie — CHANGES_REQUESTED** [review]
## Changes Required! :recycle:

### Hi @Igors78  ! :wave: :wave:

#### You have done well so far :clap: However, you did not finish all the previously requested changes. Kindly work on them and submit them for another code review.


#### Behaviour

- [x] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- I can only see the button to "Befriend", but I can't see the "Request".

- [x] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- I can't see the accept or decline buttons next to the user who sent the request.

- [x] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- You can do this by creating another view or have accept and decline buttons next to the name of the user who sent the friend request in the ALL USERS tab

- [ ] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- The timeline should only have the login user's post and his or her friend's post. At this point, the timeline shows the post of the user to who the friend request was sent before he accepts or declines it. 

### That's it, team! :raised_hands: 


__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Comment by Igors78** [comment]
Friendships requests are seen from the user profile page.
requests
the timeline shows logged_in user and friend's posts.
timeline_of_me_and_friends
Button befriend means 'request friendship'

**Review by cristianCeamatu — CHANGES_REQUESTED** [review]
## ♻️ Changes required

Hello @Igors78 ! Please make this last changes and request another review. If you need help I am ready to jump on a video call to assist.
- [ ] [Please fix] I will attach 2 screenshots of the behavior that is wrong and try to explain:
Friend1
Friend2
Mr. Robbin requested a friendship to Mr. Emanuel, Mr. Emanuel logs in and he goes on the ❗ `ALL USERS` ❗  page and he sees only the possibility to `Befriend` with Mr. Robbin (he cannot see that Mr. Robbin already made a request towards him), he does not have the possibility to `ACCEPT/DECLINE` the request but he wants to befriend with Mr. Robbin so he keeps pressing `Befriend` and keeps receiving this message back: `There is already a friend request pending`, he does not know about what request this app this talking about and gets upset, logs out and never comes back 😃 . Bottom line, you need to have the `ACCEPT/DECLINE` request options available on both `ALL USERS` and user show page. My colleague @Temmarie  mentioned this in both of her reviews.
- [ ] [Please fix] There is a button to invite the user to friendship on the users list page and on the single-user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)) You don't have the option to `Befriend` when you visit other users profile (you can `Befriend` only for the `ALL USERS` page:
Friend3
- [ ] [Optional] When you requested friendship from somebody the `Befriend` button should not appear anymore, you can replace it with a `Pending friendship` test.

If you ever need any help or notice that you are wasting too many hours on a bug don't hesitate to contact me on Slack.

Best regards,
Cristian Ceamatu
Slack: Cristian Viorel Ceamatu

**Review by oloomoses — CHANGES_REQUESTED** [review]
# Status: Changes Required
Hello team, you have done a great job implementing most of the changes :muscle: 
This project is almost ready. Before it can be approved though;
- [ ] Ensure that the user has the option to accept or reject a friend request by a click of a button. As you can see on the screenshot below, only the accept option is shown
Screenshot from 2021-02-17 14-08-06
I am sure you had this feature before. Kindly fix it before submitting another review.
#### Feel free to reach me on slack [Oloo Moses] incase of doubts or concerns

**Review by oloomoses — APPROVED** [review]
# Congratulation team! Project Approved :tada: :muscle: :100: :clap: 
Congratulation on completing this milestone. You have met the minimum requirements.
I wish you the best in the next milestone. See you at the top :rocket: 
## Happy Codding 
## [Oloo Moses]

**Review by jubaan — COMMENTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details close="close">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] Branch name does not match the milestone 6: friendshipt v2
- [ ] The branch for the requested review is merged... please open the PR.
- [ ] User model missing pending friends
- [ ] Move your model methods from the controller to the model class. The ones that apply. Example: 
```
# Friendship Model, not the Friendship Controller

def timeline_posts
  ... some code ...
end
```
- [x] Add the corresponding specs to test the new friendship feature and split test per feature.


</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [ ] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Pull request merged**